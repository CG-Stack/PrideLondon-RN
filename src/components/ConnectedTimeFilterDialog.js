// @flow
import { connect } from "react-redux";
import type { Connector } from "react-redux";
import { updateEventFilters } from "../actions/event-filters";
import { selectFilteredEvents } from "../selectors/events";
import { selectTimeFilter } from "../selectors/event-filters";
import Component from "./MultiSelectDialog";
import text from "../constants/text";

const OPTIONS = ["morning", "afternoon", "evening"];

type OwnProps = {
  onApply: () => void,
  onCancel: () => void,
  visible: boolean
};

type Props = {
  applyButtonText: string,
  onChange: (number[]) => void,
  options: string[],
  selectedIndexes: number[],
  title: string
} & OwnProps;

const mapStateToProps = state => ({
  applyButtonText: text.filterPickerApply(selectFilteredEvents(state).length),
  options: OPTIONS.map(option => text.time[option]),
  selectedIndexes: selectTimeFilter(state).map(time => OPTIONS.indexOf(time)),
  title: text.filterTimePickerTitle
});

const mapDispatchToProps = {
  onChange: selectedIndexes =>
    updateEventFilters({
      time: selectedIndexes.map(index => OPTIONS[index])
    })
};

const connector: Connector<OwnProps, Props> = connect(
  mapStateToProps,
  mapDispatchToProps
);

export default connector(Component);
