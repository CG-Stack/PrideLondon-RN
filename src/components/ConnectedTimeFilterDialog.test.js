import React from "react";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { shallow } from "enzyme";
import ConnectedTimeFilterDialog from "./ConnectedTimeFilterDialog";

const initialState = {
  events: {
    entries: [],
    assets: [],
    loading: true,
    refreshing: false
  },
  eventFilters: {
    selectedFilters: {
      categories: new Set(),
      date: null,
      timeOfDay: new Set(),
      price: new Set(),
      audience: new Set(),
      venueDetails: new Set(),
      accessibilityOptions: new Set(),
      area: new Set()
    },
    stagedFilters: {
      categories: new Set(),
      date: null,
      timeOfDay: new Set(),
      price: new Set(),
      audience: new Set(),
      venueDetails: new Set(),
      accessibilityOptions: new Set(),
      area: new Set()
    }
  }
};
const mockStore = configureStore([thunk]);

describe("ConnectedTimeFilterDialog", () => {
  it("renders connector", () => {
    const store = mockStore(initialState);
    const output = shallow(
      <ConnectedTimeFilterDialog
        store={store}
        onApply={() => {}}
        onCancel={() => {}}
        visible
      />
    );
    expect(output).toMatchSnapshot();
  });

  it("renders component", () => {
    const store = mockStore(initialState);
    const output = shallow(
      <ConnectedTimeFilterDialog
        store={store}
        onApply={() => {}}
        onCancel={() => {}}
        visible
      />
    );
    expect(output.dive()).toMatchSnapshot();
  });

  it("commits filters on apply", () => {
    const store = mockStore(initialState);
    const mockOnApply = jest.fn();
    const output = shallow(
      <ConnectedTimeFilterDialog
        store={store}
        onApply={mockOnApply}
        onCancel={() => {}}
        visible
      />
    );

    output
      .dive()
      .props()
      .onApply();
    const actions = store.getActions();

    expect(mockOnApply).toHaveBeenCalled();
    expect(actions).toEqual([{ type: "COMMIT_EVENT_FILTERS" }]);
  });

  it("clears filters on cancel", () => {
    const store = mockStore(initialState);
    const mockOnCancel = jest.fn();
    const output = shallow(
      <ConnectedTimeFilterDialog
        store={store}
        onApply={() => {}}
        onCancel={mockOnCancel}
        visible
      />
    );

    output
      .dive()
      .props()
      .onCancel();
    const actions = store.getActions();

    expect(mockOnCancel).toHaveBeenCalled();
    expect(actions).toEqual([{ type: "CLEAR_STAGED_EVENT_FILTERS" }]);
  });

  it("updates filters on change", () => {
    const store = mockStore(initialState);
    const output = shallow(
      <ConnectedTimeFilterDialog
        store={store}
        onApply={() => {}}
        onCancel={() => {}}
        visible
      />
    );

    output
      .dive()
      .find("CheckBox")
      .first()
      .props()
      .onChange();

    const actions = store.getActions();

    expect(actions[0]).toEqual(
      expect.objectContaining({ type: "STAGE_EVENT_FILTERS" })
    );
  });
});
