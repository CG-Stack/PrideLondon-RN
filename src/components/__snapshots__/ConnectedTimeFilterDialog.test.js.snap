// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectedTimeFilterDialog renders component 1`] = `
<Dialog
  applyButtonLabel="No events"
  applyButtonText="No events"
  onApply={[Function]}
  onCancel={[Function]}
  title="Select time"
  visible={true}
>
  <CheckBox
    checked={false}
    key="Morning"
    label="Morning"
    onChange={[Function]}
    style={
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      }
    }
  />
  <CheckBox
    checked={false}
    key="Afternoon"
    label="Afternoon"
    onChange={[Function]}
    style={
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      }
    }
  />
  <CheckBox
    checked={false}
    key="Evening"
    label="Evening"
    onChange={[Function]}
    style={
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      }
    }
  />
</Dialog>
`;

exports[`ConnectedTimeFilterDialog renders connector 1`] = `
<MultiSelectDialog
  applyButtonText="No events"
  onApply={[Function]}
  onCancel={[Function]}
  onChange={[Function]}
  options={
    Array [
      "Morning",
      "Afternoon",
      "Evening",
    ]
  }
  selectedIndexes={Array []}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  title="Select time"
  visible={true}
/>
`;
